examples/js/hello-world/app.js

```diff
@@ -0,0 +1,139 @@
+                bindings: [
+                    {
+                        location: 'send-button',
+                        icon: 'http://localhost:8080/static/icon.png',
+                        label: 'send hello message',
+                        call: {
+                            path: '/send-modal',
+                        },
+                    },
+                ],
+            },
+            {
+                location: '/command',
+                bindings: [
+                    {
+                        icon: 'http://localhost:8080/static/icon.png',
+                        label: 'helloworld',
+                        description: 'Hello World app',
+                        hint: '[send]',
+                        bindings: [
+                            {
+                                location: 'send',
+                                label: 'send',
+                                call: {
+                                    path: '/send',
+                                },
+                            },
+                        ],
+                    },
+                ],
+            },
+        ],
+    });
+});
+
+app.post(['/send/form', '/send-modal/submit'], (req, res) => {
+    res.json({
+        type: 'form',
+        form: {
+            title: 'Hello, world!',
+            icon: 'http://localhost:8080/static/icon.png',
+            fields: [
+                {
+                    type: 'text',
+                    name: 'message',
+                    label: 'message',
+                },
+            ],
+            call: {
+                path: '/send',
+            },
+        },
+    });
+});
+
+app.get('/static/icon.png', (req, res) => {
+    res.sendFile(__dirname + '/icon.png');
+});
+
+app.use(express.json());
+
+app.post('/send/submit', (req, res) => {
+    const call = req.body;
+
+    let message = 'Hello, world!';
+    const user_message = call.values.message;
+    if (user_message) {
+        message += ' ...and ' + user_message + '!';
+    }
+
+    const users = [
+        call.context.bot_user_id,
+        call.context.acting_user_id,
+    ];
+
+    // Use the app bot to do API calls
+    const options = {
+        method: 'POST',
+        headers: {
+            Authorization: 'BEARER ' + call.context.bot_access_token,
+            'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(users),
+    };
+
+    // Get the DM channel between the user and the bot
+    const mattermost_site_url = call.context.mattermost_site_url;
+
+    fetch(mattermost_site_url + '/api/v4/channels/direct', options).
+        then((mm_res) => mm_res.json()).
+        then((channel) => {
+            const post = {
+                channel_id: channel.id,
+                message,
+            };
+
+            // Create a post
+            options.body = JSON.stringify(post);
+
+            fetch(mattermost_site_url + '/api/v4/posts', options);
+        });
```

Comment:
Would you be opposed to using `async/await` here?

```suggestion
    const channel = await fetch(mattermost_site_url + '/api/v4/channels/direct', options).then((mm_res) => mm_res.json());
        
    const post = {
        channel_id: channel.id,
        message,
    };

    // Create a post
    options.body = JSON.stringify(post);

    fetch(mattermost_site_url + '/api/v4/posts', options);
```

Reply:
@mickmister That snipped errors when I run it locally:
```
SyntaxError: await is only valid in async functions and the top level bodies of modules
```

Is `fetch` any async function?

Reply:
Yes `fetch` returns a promise, making it an async function as far as the caller is concerned. This error above occurs because we are attempting to use the `await` keyword inside of a function that was not declared as `async`. This change needs to be made:

```diff
- app.post('/send/submit', (req, res) => {
+ app.post('/send/submit', async (req, res) => {
```

Reply:
Any function that is `then`-able (meaning it has a `then` method that you can chain) returns a promise. Therefore, any function that is `then`-able can alternatively be called with `await`, instead of using `.then()`.